lib/codecs/src/encoding/format/protobuf.rs

```diff
@@ -0,0 +1,342 @@
+        let message = encode_message(
+            &test_message_descriptor("RepeatedPrimitive"),
+            Value::Object(BTreeMap::from([(
+                "numbers".into(),
+                Value::Array(vec![
+                    Value::Integer(8),
+                    Value::Integer(6),
+                    Value::Integer(4),
+                ]),
+            )])),
+        )
+        .unwrap();
+        let list = mfield!(message, "numbers").as_list().unwrap().to_vec();
+        assert!(list.len() == 3);
+        assert!(list[0].as_i64() == Some(8));
+        assert!(list[1].as_i64() == Some(6));
+        assert!(list[2].as_i64() == Some(4));
+    }
+
+    #[test]
+    fn test_encode_repeated_message() {
+        let message = encode_message(
+            &test_message_descriptor("RepeatedMessage"),
+            Value::Object(BTreeMap::from([(
+                "messages".into(),
+                Value::Array(vec![
+                    Value::Object(BTreeMap::from([(
+                        "text".into(),
+                        Value::Bytes(Bytes::from("vector")),
+                    )])),
+                    Value::Object(BTreeMap::from([("index".into(), Value::Integer(4444))])),
+                    Value::Object(BTreeMap::from([
+                        ("text".into(), Value::Bytes(Bytes::from("protobuf"))),
+                        ("index".into(), Value::Integer(1)),
+                    ])),
+                ]),
+            )])),
+        )
+        .unwrap();
+        let list = mfield!(message, "messages").as_list().unwrap().to_vec();
+        assert!(list.len() == 3);
+        assert!(mfield!(list[0].as_message().unwrap(), "text").as_str() == Some("vector"));
+        assert!(!list[0].as_message().unwrap().has_field_by_name("index"));
+        assert!(!list[1].as_message().unwrap().has_field_by_name("t4ext"));
+        assert!(mfield!(list[1].as_message().unwrap(), "index").as_u32() == Some(4444));
+        assert!(mfield!(list[2].as_message().unwrap(), "text").as_str() == Some("protobuf"));
+        assert!(mfield!(list[2].as_message().unwrap(), "index").as_u32() == Some(1));
+    }
+
+    #[test]
+    fn test_encode_decoding_protobuf_test_data() {
+        let test_data_dir = PathBuf::from(std::env::var_os("CARGO_MANIFEST_DIR").unwrap())
+            .join("tests/data/decoding/protobuf");
+
+        // test_protobuf (proto2)
+        let descriptor_set_path = test_data_dir.join("test_protobuf.desc");
+        let message_type = "test_protobuf.Person";
+        let message_descriptor =
+            get_message_descriptor(&descriptor_set_path, message_type).unwrap();
+        // just check for the side-effect of success
+        encode_message(
+            &message_descriptor,
+            Value::Object(BTreeMap::from([
+                ("name".into(), Value::Bytes(Bytes::from("gina"))),
+                ("id".into(), Value::Integer(9271)),
+            ])),
+        )
+        .unwrap();
```

Comment:
💬 Let's extract this into a function since it's repeated below.

Reply:
Extracted to `run_encoding_on_decoding_test_data`.